### PR TITLE
Add: `MelSpectrogram` layer

### PR DESCRIPTION
**Pull request by awsaf49**
`MelSpectrogram` layer was a long overdue. It got halted during the keras v2 to v3 port and was stuck for a while due to some missing dependencies. This PR will add this layer along with it's dependencies (`linear_to_mel_weight_matrix`, `frame`, etc).

* issue (keras-v3): https://github.com/keras-team/keras/issues/18405
* issue (keras-v2): https://github.com/keras-team/tf-keras/issues/55
* PR (keras-v2): https://github.com/keras-team/keras/pull/17717

### CODE REVIEW

**Comment by awsaf49** [comment]
cc: @fchollet  kindly need your review on this, specifically I'm not sure how this layer will fit to current keras v3. For now, I'm keeping everything in `layers/preprocessing/audio_preprocessing.py`. 

Also, while porting from TensorFlow I skipped dynamic shape parts as I wasn't sure how they translate to keras 3. So need to do some tests.

**Comment by awsaf49** [comment]
Visual results seem fine,
image

```py
import librosa
import matplotlib.pyplot as plt
import librosa.display

num_mel_bins = 256
fft_stride = 128
num_fft_bins = 1024
fmin = 512
fmax = 8000

y, sr = librosa.load("/cardinal-37075.mp3", sr=None)
audio = keras.ops.convert_to_tensor(y)

spec = keras.layers.MelSpectrogram(
    num_mel_bins=num_mel_bins,
    sampling_rate=sr,
    fft_stride=fft_stride,
    num_fft_bins=num_fft_bins,
    min_freq=fmin,
    max_freq=fmax,
)(audio)
spec = keras.ops.convert_to_numpy(spec)


# display mel spectrogram with librosa
plt.figure(figsize=(10, 3 * 2))

plt.subplot(211)
librosa.display.specshow(
    spec,
    hop_length=fft_stride,
    sr=sr,
    n_fft=num_fft_bins,
    y_axis="mel",
    x_axis="time",
    fmin=fmin,
    fmax=fmax,
)
plt.xlabel(None)

# plot the waveform
plt.subplot(212)
librosa.display.waveshow(y, sr=sr)
plt.xlim([0, len(y)/sr]) # set x-axis limits

plt.tight_layout()
plt.show()
```

File: keras/layers/preprocessing/audio_preprocessing.py [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+    Output shape:
+        2D (unbatched) or 3D (batched) tensor with
+        shape:`(..., num_mel_bins, time)`.
+
+    Args:
+        num_fft_bins: Integer, size of the FFT window.
+        fft_stride: Integer, number of samples between successive STFT columns.
+        window_size: Integer, size of the window used for applying `window_fn`
+            to each audio frame. If `None`, defaults to `num_fft_bins`.
+        window_fn: String, name of the window function to use.
```
Comment (fchollet) [inline keras/layers/preprocessing/audio_preprocessing.py:417]: Argument descriptions like this should list the possible values.

File: keras/layers/preprocessing/audio_preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+            -0.7777866 , -15.290524  ]]], dtype=float32)>
+
+    Input shape:
+        1D (unbatched) or 2D (batched) tensor with shape:`(..., samples)`.
+
+    Output shape:
+        2D (unbatched) or 3D (batched) tensor with
+        shape:`(..., num_mel_bins, time)`.
+
+    Args:
```
Comment (fchollet) [inline keras/layers/preprocessing/audio_preprocessing.py:412]: We need to make sure that the argument names are consistent with our existing FFT ops.

File: keras/layers/preprocessing/audio_preprocessing.py [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+            `10 * log10(S / ref_power)`.
+        min_power: Float, minimum value for power and `ref_power`.
+    """
+
+    def __init__(
+        self,
+        num_fft_bins=2048,
+        fft_stride=512,
+        window_size=None,
+        window="hann",
```
Comment (fchollet) [inline keras/layers/preprocessing/audio_preprocessing.py:437]: There's a mismatch with the name in the description

File: keras/layers/preprocessing/audio_preprocessing.py [hunk 4]
```diff
@@ -0,0 +1,548 @@
...
+            outputs = self.dbscale(outputs)
+        # swap time & freq axis to have shape of (..., num_mel_bins, time)
+        outputs = ops.swapaxes(outputs, -1, -2)
+        outputs = ops.cast(outputs, self.compute_dtype)
+        return outputs
+
+    def spectrogram(self, inputs):
+        real, imag = ops.stft(
+            inputs,
+            sequence_length=self.window_size,
```
Comment (fchollet) [inline keras/layers/preprocessing/audio_preprocessing.py:478]: We need to use the same argument names for consistency.

File: keras/layers/preprocessing/audio_preprocessing.py [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+            fft_length=self.num_fft_bins,
+            window=self.window,
+            center=True,
+        )
+        # abs of complex  = sqrt(real^2 + imag^2)
+        spec = ops.sqrt(ops.add(ops.square(real), ops.square(imag)))
+        spec = ops.power(spec, self.mag_exp)
+        return spec
+
+    def melscale(self, inputs):
```
Comment (fchollet) [inline keras/layers/preprocessing/audio_preprocessing.py:489]: Use private methods.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! It seems like there's a lot of interest in having such a layer available, so we can definitely add it.

File: keras/layers/preprocessing/audio_preprocessing.py [hunk 6]
```diff
@@ -0,0 +1,548 @@
...
+
+    a, b = ops.while_loop(
+        lambda _, b: ops.greater(b, ops.zeros_like(b)),
+        lambda a, b: [b, ops.mod(a, b)],
+        [a, b],
+    )
+    return a
+
+
+def frame(
```
Comment (james77777778) [inline keras/layers/preprocessing/audio_preprocessing.py:33]: Hi @awsaf49 
I believe we already have `ops.math.extract_sequences` to handle the framing.
We should be able to extend the current API to support `pad_end`, `pad_value` and `axis` if needed.

https://github.com/keras-team/keras/blob/master/keras/ops/math.py#L277

Comment (awsaf49) [reply keras/layers/preprocessing/audio_preprocessing.py:33]: Thanks, I'll check it out too

Comment (awsaf49) [reply keras/layers/preprocessing/audio_preprocessing.py]: @james77777778 thanks for the suggestion, I've replaced `frame` with `ops.math.extract_sequences` as for melspec layer I don't need the `pad` option. Feel free to extend it for padding.

File: keras/layers/preprocessing/audio_preprocessing.py [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+# mel spectrum constants.
+_MEL_BREAK_FREQUENCY_HERTZ = 700.0
+_MEL_HIGH_FREQUENCY_Q = 1127.0
+
+
+def _mel_to_hertz(mel_values, name=None):
+    """Converts frequencies in `mel_values` from the mel scale to linear scale.
+
+    Args:
+    mel_values: A `Tensor` of frequencies in the mel scale.
```
Comment (fchollet) [inline keras/layers/preprocessing/audio_preprocessing.py:16]: - Add indent for arguments
- There's no cross-backend `Tensor` class, so just say "a tensor"

File: keras/layers/preprocessing/audio_preprocessing.py [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+    name: An optional name for the operation.
+
+    Returns:
+    A `Tensor` of the same shape and type as `mel_values` containing linear
```
Comment (fchollet) [inline keras/layers/preprocessing/audio_preprocessing.py:20]: - Add indent
- Just say "A tensor"

File: keras/layers/preprocessing/audio_preprocessing.py [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+    return _MEL_BREAK_FREQUENCY_HERTZ * (
+        ops.exp(mel_values / _MEL_HIGH_FREQUENCY_Q) - 1.0
+    )
+
+
+def _hertz_to_mel(frequencies_hertz, name=None):
+    """Converts frequencies in `frequencies_hertz` in Hertz to the mel scale.
+
+    Args:
+    frequencies_hertz: A `Tensor` of frequencies in Hertz.
```
Comment (fchollet) [inline keras/layers/preprocessing/audio_preprocessing.py:33]: Same comments

File: keras/layers/preprocessing/audio_preprocessing.py [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+    name: An optional name for the operation.
+
+    Returns:
+    A `Tensor` of the same shape and type of `frequencies_hertz` containing
```
Comment (fchollet) [inline keras/layers/preprocessing/audio_preprocessing.py:37]: Same comments

File: keras/layers/preprocessing/audio_preprocessing.py [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+    Returns a weight matrix that can be used to re-weight a `Tensor` containing
+    `num_spectrogram_bins` linearly sampled frequency information from
+    `[0, sampling_rate / 2]` into `num_mel_bins` frequency information from
+    `[lower_edge_hertz, upper_edge_hertz]` on the [mel scale][mel].
+
+    This function follows the [Hidden Markov Model Toolkit
+    (HTK)](http://htk.eng.cam.ac.uk/) convention, defining the mel scale in
+    terms of a frequency in hertz according to the following formula:
+
+    `mel(f) = 2595 * log10( 1 + f/700)`
```
Comment (fchollet) [inline keras/layers/preprocessing/audio_preprocessing.py:66]: You can use a code block here (3 backticks)

File: keras/layers/preprocessing/audio_preprocessing.py [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+
+def linear_to_mel_weight_matrix(
+    num_mel_bins=20,
+    num_spectrogram_bins=129,
+    sampling_rate=8000,
+    lower_edge_hertz=125.0,
+    upper_edge_hertz=3800.0,
+    dtype="float32",
+):
+    r"""Returns a matrix to warp linear scale spectrograms
```
Comment (fchollet) [inline keras/layers/preprocessing/audio_preprocessing.py:54]: The one-line description should fit on one line. Also no need for the leading `r`

File: keras/layers/preprocessing/audio_preprocessing.py [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+        to the [mel scale][mel].
+
+    Returns a weight matrix that can be used to re-weight a `Tensor` containing
+    `num_spectrogram_bins` linearly sampled frequency information from
+    `[0, sampling_rate / 2]` into `num_mel_bins` frequency information from
+    `[lower_edge_hertz, upper_edge_hertz]` on the [mel scale][mel].
+
+    This function follows the [Hidden Markov Model Toolkit
```
Comment (fchollet) [inline keras/layers/preprocessing/audio_preprocessing.py:62]: Links should be on a single line, or broken after the `(`

File: keras/layers/preprocessing/audio_preprocessing.py [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+            included in the mel spectrum. This corresponds to the lower edge of
+            the lowest triangular band.
+        upper_edge_hertz: Python float. The desired top edge of the highest
+            frequency band.
+        dtype: The `DType` of the result matrix. Must be a floating point type.
+
+    Returns:
+        A `Tensor` of shape `[num_spectrogram_bins, num_mel_bins]`.
+
+    [mel]: https://en.wikipedia.org/wiki/Mel_scale
```
Comment (fchollet) [inline keras/layers/preprocessing/audio_preprocessing.py:111]: You can do:

```
References:

    - [Mel scale (Wikipedia)](https://en.wikipedia.org/wiki/Mel_scale)
```

**Review by fchollet — COMMENTED** [review]
Looking good! Please add a set of unit tests as well.

**Comment by awsaf49** [comment]
> Looking good! Please add a set of unit tests as well.

Thanks. I will add the unit test soon.

**Comment by awsaf49** [comment]
@fchollet, while performing correctness tests, I noticed that the result from `keras` matches with `torchaudio` but doesn't match with `librosa`, even when the parameters are the same. However, visually, they have similar shape results. Here is a numerical comparison:

## **Keras**:

**Code:**
```python
spec_keras = keras.layers.MelSpectrogram(
    num_mel_bins=num_mel_bins,
    sampling_rate=sr,
    sequence_stride=fft_stride,
    fft_length=num_fft_bins,
    min_freq=fmin,
    max_freq=fmax,
    ref_power=1.0,
)(y)
spec_keras = keras.ops.convert_to_numpy(spec_keras)
```
**Output**:
```python
array([[-38.097645, -38.097645, -38.097645, ..., -31.054651, -31.315052,
         -32.428814],
        [-38.097645, -38.097645, -38.097645, ..., -38.097645, -38.097645,
         -38.097645],
        [-38.097645, -38.097645, -38.097645, ..., -33.124752, -38.097645,
         -38.06842 ],
        ...,
        [-38.097645, -38.097645, -38.097645, ..., -38.097645, -38.097645,
         -38.097645],
        [-38.097645, -38.097645, -38.097645, ..., -38.097645, -38.097645,
         -38.097645],
        [-38.097645, -38.097645, -38.097645, ..., -38.097645, -38.097645,
         -38.097645]], dtype=float32),
```

## **Librosa**:

**Code**:
```python
spec_librosa = librosa.feature.melspectrogram(y=y, sr=sr, n_fft=num_fft_bins,
                                      hop_length=fft_stride, n_mels=num_mel_bins,
                                      fmin=fmin, fmax=fmax)
spec_librosa = librosa.power_to_db(spec_librosa, ref=1.0, top_db=80.0)
```

**Output**:
```python
array([[-53.241608, -53.241608, -53.241608, ..., -41.145687, -42.29181 ,
         -46.312866],
        [-53.241608, -53.241608, -53.241608, ..., -48.96924 , -53.241608,
         -53.241608],
        [-53.241608, -53.241608, -53.241608, ..., -43.543972, -53.241608,
         -51.195637],
        ...,
        [-53.241608, -53.241608, -53.241608, ..., -53.241608, -53.241608,
         -53.241608],
        [-53.241608, -53.241608, -53.241608, ..., -53.241608, -53.241608,
         -53.241608],
        [-53.241608, -53.241608, -53.241608, ..., -53.241608, -53.241608,
         -53.241608]], dtype=float32)
```

## **Torchaudio**:

**Code**:
```python
mel_spec_transform = torchaudio.transforms.MelSpectrogram(
    sample_rate=sr,
    n_fft=num_fft_bins,
    hop_length=fft_stride,
    n_mels=num_mel_bins,
    f_min=fmin,
    f_max=fmax
)
spec_torch = mel_spec_transform(y)
spec_torch = torchaudio.transforms.AmplitudeToDB(top_db=80.0)(spec_torch).numpy()
```

**Output**
```python
array([[-38.098427, -38.098427, -38.098427, ..., -31.064793, -31.325191,
         -32.438957],
        [-38.098427, -38.098427, -38.098427, ..., -38.098427, -38.098427,
         -38.098427],
        [-38.098427, -38.098427, -38.098427, ..., -33.12678 , -38.098427,
         -38.070446],
        ...,
        [-38.098427, -38.098427, -38.098427, ..., -38.098427, -38.098427,
         -38.098427],
        [-38.098427, -38.098427, -38.098427, ..., -38.098427, -38.098427,
         -38.098427],
        [-38.098427, -38.098427, -38.098427, ..., -38.098427, -38.098427,
         -38.098427]], dtype=float32))
```

**Comment by fchollet** [comment]
This appears to be correct and normal: https://github.com/pytorch/audio/issues/1058

> I found librosa use slaney normalization for the mel-filterbank creation as the default, while torchaudio is no normalization by default. 

One would one go about adding slaney normalization in Keras? Is it possible for the user to add it externally from the layer?

**Comment by awsaf49** [comment]
> This appears to be correct and normal: [pytorch/audio#1058](https://github.com/pytorch/audio/issues/1058)
> 
> > I found librosa use slaney normalization for the mel-filterbank creation as the default, while torchaudio is no normalization by default.
> 
> How would one go about adding slaney normalization in Keras? Is it possible for the user to add it externally from the layer?

I tried to implement "slaney" normalization in keras from librosa [ref](https://librosa.org/doc/main/_modules/librosa/filters.html#mel) but the values still doesn't match.

```py
    if norm == "slaney":
        # Slaney-style mel is scaled to be approx constant energy per channel
        mel_f = ops.math.extract_sequences(
            ops.linspace(
                _hertz_to_mel(lower_edge_hertz),
                _hertz_to_mel(upper_edge_hertz),
                num_mel_bins + 2,
            ),
            sequence_length=1,
            sequence_stride=1,
        )
        enorm = 2.0 / (mel_f[2 : num_mel_bins + 2] - mel_f[:num_mel_bins])
        mel_weights_matrix *= ops.transpose(enorm)
    elif norm is not None:
        raise NotImplementedError(f"Unsupported norm={norm}")
```

## **output**:

**Before**:
```py
array([[-38.097645, -38.097645, -38.097645, ..., -31.054651, -31.315052,
         -32.428814],
        [-38.097645, -38.097645, -38.097645, ..., -38.097645, -38.097645,
         -38.097645],
        [-38.097645, -38.097645, -38.097645, ..., -33.124752, -38.097645,
         -38.06842 ],
        ...,
        [-38.097645, -38.097645, -38.097645, ..., -38.097645, -38.097645,
         -38.097645],
        [-38.097645, -38.097645, -38.097645, ..., -38.097645, -38.097645,
         -38.097645],
        [-38.097645, -38.097645, -38.097645, ..., -38.097645, -38.097645,
         -38.097645]], dtype=float32),
```

**After**:
```py
array([[-47.464535, -47.464535, -47.464535, ..., -40.421543, -40.68195 ,
         -41.795708],
        [-47.464535, -47.464535, -47.464535, ..., -47.464535, -47.464535,
         -47.464535],
        [-47.464535, -47.464535, -47.464535, ..., -42.491627, -47.464535,
         -47.435295],
        ...,
        [-47.464535, -47.464535, -47.464535, ..., -47.464535, -47.464535,
         -47.464535],
        [-47.464535, -47.464535, -47.464535, ..., -47.464535, -47.464535,
         -47.464535],
        [-47.464535, -47.464535, -47.464535, ..., -47.464535, -47.464535,
         -47.464535]], dtype=float32)
```

**LIbrosa**:
```py
array([[-53.241608, -53.241608, -53.241608, ..., -41.145687, -42.29181 ,
         -46.312866],
        [-53.241608, -53.241608, -53.241608, ..., -48.96924 , -53.241608,
         -53.241608],
        [-53.241608, -53.241608, -53.241608, ..., -43.543972, -53.241608,
         -51.195637],
        ...,
        [-53.241608, -53.241608, -53.241608, ..., -53.241608, -53.241608,
         -53.241608],
        [-53.241608, -53.241608, -53.241608, ..., -53.241608, -53.241608,
         -53.241608],
        [-53.241608, -53.241608, -53.241608, ..., -53.241608, -53.241608,
         -53.241608]], dtype=float32)
```

**Comment by fchollet** [comment]
Did you try to compare the output of `mel_weights_matrix` side by side between the two implementations?

If it matches, then this implication is that there's an additional difference with Librosa.

In any case, I think that the match with torch audio is sufficient validation of the correctness of the implementation.

**Comment by awsaf49** [comment]
@fchollet We can use an external mel matrix like this if we want to use `norm="slaney"`. I've checked the result, and it matches with `librosa`, which means a mismatch is happening in the mel weights. 
```python
class MelSpectrogram(keras.layers.MelSpectrogram):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)

    def _melscale(self, inputs):
        weights = librosa.filters.mel(
            sr=self.sampling_rate, n_fft=self.fft_length,
            n_mels=self.num_mel_bins, fmin=self.min_freq, 
            fmax=self.max_freq, htk=True, norm="slaney",
        )
        weights = keras.backend.convert_to_tensor(weights.T)
        return keras.backend.tensordot(inputs, weights, axes=1)
```

I've tried comparing the mel matrix from Keras and Librosa side by side with the following code:

```python
matrix_librosa = librosa.filters.mel(
                sr=sr, n_fft=num_fft_bins, n_mels=num_mel_bins,
                fmin=fmin, fmax=fmax, norm="slaney", htk=True,
            )

matrix_keras = linear_to_mel_weight_matrix(
    num_mel_bins=num_mel_bins,
    num_spectrogram_bins=num_fft_bins // 2 + 1,
    sampling_rate=sr,
    lower_edge_hertz=fmin,
    upper_edge_hertz=fmax,
    norm="slaney",
).T


(matrix_keras - matrix_librosa).sum()
```

When `htk=True` and `norm=None`, the Librosa mel matrix matches exactly with Keras, and the final output from mel-spec layers also matches. Mismatches happen when we set `norm="slaney"` in Keras and Librosa; then, the mel matrix doesn't match.

**So, If we want an identical result across `librosa`, `keras`, and `torchaudio`, all we need to do is set `htk=True` and `norm=None`.**

**Comment by awsaf49** [comment]
@fchollet should I add `correctness` test with `torchaudio`? 

Also do we need `tf_data` compatibility? I tried adding `tf_data_compatibility` test. But it fails.

**code**:
```py
input_data = np.random.random((2, 8000))
layer = keras.layers.MelSpectrogram(num_mel_bins=80,
                                  sampling_rate=8000,
                                  sequence_stride=128,
                                  fft_length=2048)
ds = tf.data.Dataset.from_tensor_slices(input_data).batch(2)

ds = ds.map(layer)
```
**error**:
```shell
    NotImplementedError: Exception encountered when calling MelSpectrogram.call().
    
    Cannot convert a symbolic tf.Tensor (Cast:0) to a numpy array. This error may indicate that you're trying to pass a Tensor to a NumPy call, which is not supported.
    
    Arguments received by MelSpectrogram.call():
      • inputs=<tf.Tensor 'Cast:0' shape=(None, 8000) dtype=float32>
```
But when I try this, it runs fine
**code**:
```py
layer(next(iter(ds.take(1))))
```

**Comment by awsaf49** [comment]
> As for whether we should do it -- how do you think the layer will be used? Inside a model? Externally?

Using this layer in a model is definitely the best choice, as for later usage, all we have to do is load the model and then pass the audio directly to the model. If we use this layer in data pipeline, then we have to process the audio separately with proper hyper parameters before passing audio to the model. I think [**YAMNet**](https://www.tensorflow.org/hub/tutorials/yamnet) in TensorFlow Hub does the same."

**Comment by fchollet** [comment]
Ok -- it's definitely better to make it work with tf.data even if that's not the main use case, then. If you're running into major blockers, then we can revert and merge a version that only works with tf.data when using the TF backend. That's ok.

**Comment by awsaf49** [comment]
@fchollet I've added support for `tf.data` but as it required using `self.backend` instead of `keras.ops`, I had to move the independent functions into `MelSpectrogram` layer. I've checked it, it runs fine in both model and tf.data.

**Comment by fchollet** [comment]
Right -- if you want tf.data support with a backend other than TF, all ops applied by the layer must go through `self.backend`. Anything else won't work.

**Review by fchollet — APPROVED** [review]
LGTM. Nice work -- thank you for the useful contribution!

**Comment by daniel-deychakiwsky** [comment]
Is this layer compatible/convertible to TFLite?

**Pull request merged**